flink-core/src/main/java/org/apache/flink/core/fs/SafeFileSystemFactory.java

```diff
@@ -18,29 +18,52 @@
 package org.apache.flink.core.fs;
 
 import org.apache.flink.configuration.Configuration;
+import org.apache.flink.util.FatalExitExceptionHandler;
 import org.apache.flink.util.TemporaryClassLoaderContext;
 import org.apache.flink.util.WrappingProxy;
+import org.apache.flink.util.concurrent.ExecutorThreadFactory;
 
 import java.io.IOException;
 import java.net.URI;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
 
 /**
- * A wrapper around {@link FileSystemFactory} that ensures the plugin classloader is used for all
- * {@link FileSystem} operations.
+ * A wrapper around {@link FileSystemFactory} that ensures the same classloader is used for all
+ * {@link FileSystem} operations and that we don't leak classloader via inherited {@link
+ * java.security.ProtectionDomain protection domains} when new threads that might outlive the
+ * classloader are spawned during {@link FileSystem} initialization.
  */
-public class PluginFileSystemFactory implements FileSystemFactory {
+public class SafeFileSystemFactory implements FileSystemFactory {
```

Comment:
Note sure if the rename is necessary. I'd argue that your change also works for a `PluginFileSystemFactory`.